### PR TITLE
Deduplicate scan results

### PR DESCRIPTION
**Pull request by oheger-bosch**
This PR prevents that duplicate scan results are stored if multiple scanner workers are run in parallel. It contains a migration that cleans up existing duplicates and introduces unique constraints. Note that the migration will take a while on larger databases (50 minutes in our setup).

### CODE REVIEW

File: dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+-- Update the value of the sequence for this table, since this was not done by the previous INSERT statement.
+SELECT setval(
+  'scan_results_additional_data_id_seq',
+  (SELECT MAX(id) FROM scan_results_additional_data),
+  TRUE
+);
+
+ALTER TABLE scan_results
+  ADD COLUMN additional_data_id bigint REFERENCES scan_results_additional_data;
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql:26]: Which kind of issues do you see? You are right, that automatic deletion via CASCADE will not not work, so deleting scan results will be more complex. The proposed alternative with the hash column will instead increase the complexity when creating new entities and introduce some redundancy; so I am unsure about the benefit.

File: dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql [hunk 2]
```diff
@@ -0,0 +1,156 @@
+-- This migration script deals with scan results that have been duplicated due to a race condition when
+-- multiple ORT runs process the same packages concurrently. To better deal with duplicates, it extracts the
+-- unstructured additional data into a dedicated table which gets referenced from scan_results. Note that this
+-- property is currently only set by FossID; for results from ScanCode, it is empty. By making use of this property,
+-- it is now possible to again reintroduce unique constraints on the scan_results table, which had been removed in
+-- migration 47.
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql:6]: Yes, this is correct. The additional data is part of the identifying properties of a scan result; but AFAIK, JSONB columns or text columns with arbitrary length will not work with an index.

File: workers/scanner/src/main/kotlin/scanner/OrtServerScanResultStorage.kt [hunk 3]
```diff
@@ -108,75 +108,84 @@ class OrtServerScanResultStorage(
...
-                this.startTime = scanResult.summary.startTime.toKotlinInstant()
-                this.endTime = scanResult.summary.endTime.toKotlinInstant()
-            }
+            val resultDao = findExistingScanResult(scanResult) ?: createNewScanResult(scanResult, provenance)
 
-            scanResult.summary.issues.forEach {
-                ScanSummariesIssuesDao.createByIssue(summaryDao.id.value, it.mapToModel())
-            }
+            associateScanResultWithScannerRun(resultDao)
+        }
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply workers/scanner/src/main/kotlin/scanner/OrtServerScanResultStorage.kt:114]: There is a default value for the maximum attempts, which I believe is 3. Which value do you propose?

File: dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql:26]: Another alternative could be using a unique constraint with a hash function as described [here](https://dba.stackexchange.com/questions/334939/computing-hashes-in-postgresql-for-jsonb-columns). I would need to experiment with that.

[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql]: Yes, the `jsonb_hash_extended` looks promising to me.

File: workers/scanner/src/main/kotlin/scanner/OrtServerScanResultStorage.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply workers/scanner/src/main/kotlin/scanner/OrtServerScanResultStorage.kt:122]: Ok, 3 sounds fine.

File: dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply dao/src/main/resources/db/migration/V87__deduplicateScanResults.sql]: This approach has been implemented. It actually speeds up the migration significantly; the execution time on our environment was reduced from ~53 minutes to ~7 minutes.

**Review by mnonnenmacher — APPROVED** [review]

**Pull request merged**